src/scope.rs

```diff
@@ -174,18 +188,287 @@ impl Pattern {
+            .into_iter()
+            .filter_map(|d| match d {
+                Direction::Left => Some(Short::O),
+                Direction::Right => Some(Short::I),
+                _ => None,
+            })
+            .collect();
+        Some(short)
+    }
+
+    /// Get a Simplicity expression that returns a value of the shape of the target.
+    /// The expression takes a Simplicity value of the shape of this pattern as input.
+    fn get_translation_node(&self, target: &Self) -> Result<TranslationNode, Identifier> {
+        let mut output = vec![];
+        for data in target.to_base().post_order_iter() {
+            match data.node {
+                Pattern::Identifier(i) => {
+                    let path = self.get_path(i).ok_or(i.clone())?;
+                    output.push(TranslationNode::Short(path));
+                }
+                Pattern::Ignore => {
+                    output.push(TranslationNode::Unit);
+                }
+                Pattern::Product(..) => {
+                    let node_r = output.pop().unwrap();
+                    let node_l = output.pop().unwrap();
+                    output.push(TranslationNode::Pair(Arc::new(node_l), Arc::new(node_r)));
+                }
+                Pattern::Array(_) => unreachable!(),
+            }
+        }
+        debug_assert!(output.len() == 1);
+        Ok(output.pop().unwrap())
+    }
+
+    /// Get a compressed Simplicity expression that returns a value of the shape of the target.
+    /// The expression takes a Simplicity value of the shape of this pattern as input.
+    pub fn translate(&self, other: &Self) -> Result<ProgNode, Identifier> {
+        let node = self.get_translation_node(other)?;
+        let compressed = Arc::new(node.compress());
+        Ok(compressed.to_expr())
+    }
+}
+
+/// Abbreviation for the take or drop combinator.
+///
+/// Sequences of [`Short::O`] and [`Short::I`] are followed by an implicit `iden` combinator.
+/// Because `iden` always follows, it doesn't make sense to explicitly add it in code.
+/// The sequence `[Short::O, Short::I]` stands for `take drop iden`, and so on.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Short {
+    /// Take
+    O,
+    /// Drop
+    I,
+}
+
+/// Simplicity expression that translates a nested Simplicity product value into another.
+///
+/// Only a small set of combinators is needed for this operation.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum TranslationNode {
+    Unit,
+    Short(Vec<Short>),
+    Pair(Arc<Self>, Arc<Self>),
+}
+
+impl<'a> TreeLike for &'a TranslationNode {
+    fn as_node(&self) -> Tree<Self> {
+        match self {
+            TranslationNode::Unit | TranslationNode::Short(_) => Tree::Nullary,
+            TranslationNode::Pair(l, r) => Tree::Binary(l, r),
+        }
+    }
+}
+
+impl TranslationNode {
+    /// Eliminate an unnecessary pair if possible.
+    /// Return the minimized node.
+    fn eliminate_pair(&self) -> Option<Self> {
```

Comment:
In 44ccbdc6ca21f1113042aa53f37d84ef2bc6464e:

This function seems like it's the heart of this commit but I'm struggling to understand it. Maybe we should talk on the phone about it.